tsn/src/bin/throughput.rs

```diff
@@ -0,0 +1,677 @@
+use bincode;
+use clap::{Arg, Command as ClapCommand};
+use nix::sys::time::TimeSpec;
+use nix::sys::time::TimeValLike;
+use nix::time::clock_gettime;
+use nix::time::ClockId;
+use serde::{Deserialize, Serialize};
+use signal_hook::{consts::SIGINT, iterator::Signals};
+use socket;
+use std::io::Error;
+use std::io::{self, Write};
+use std::mem;
+use std::str::FromStr;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::vec::Vec;
+use std::{thread, time::Duration};
+const VLAN_ID_PERF: u32 = 10;
+const VLAN_PRI_PERF: u32 = 3;
+const ETHERTYPE_PERF: u16 = 0x1337;
+static RUNNING: AtomicBool = AtomicBool::new(true);
+const TIMEOUT_SEC: u32 = 1;
+
+#[derive(Clone, Copy, Serialize, Deserialize, Debug)]
+enum PerfOpcode {
+    PerfReqStart = 0x00,
+    PerfReqEnd = 0x01,
+    PerfResStart = 0x20,
+    PerfResEnd = 0x21,
+    PerfData = 0x30,
+    PerfReqResult = 0x40,
+    PerfResResult = 0x41,
+}
+
+impl From<u8> for PerfOpcode {
+    fn from(value: u8) -> Self {
+        match value {
+            0x00 => PerfOpcode::PerfReqStart,
+            0x01 => PerfOpcode::PerfReqEnd,
+            0x20 => PerfOpcode::PerfResStart,
+            0x21 => PerfOpcode::PerfResEnd,
+            0x30 => PerfOpcode::PerfData,
+            0x40 => PerfOpcode::PerfReqResult,
+            0x41 => PerfOpcode::PerfResResult,
+            _ => panic!("Invalid opcode value"),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct Ethernet {
+    dest: [u8; 6],
+    src: [u8; 6],
+    ether_type: u16,
+}
+#[derive(Serialize, Deserialize)]
+struct PktInfo {
+    id: u32,
+    op: u8,
+}
+
+#[derive(Serialize, Deserialize)]
+struct PktPerfResult {
+    pkt_count: u64,
+    pkt_size: u64,
+    elapsed_sec: i64,
+    elapsed_nsec: i64,
+}
+
+struct Statistics {
+    pkt_count: u64,
+    total_bytes: u64,
+    last_id: u32,
+    running: bool,
+}
+
+static mut STATS: Statistics = Statistics {
+    pkt_count: 0,
+    total_bytes: 0,
+    last_id: 0,
+    running: true,
+};
+
+static mut SOCK: tsn::TsnSocket = tsn::TsnSocket {
+    fd: 0,
+    ifname: String::new(),
+    vlanid: 0,
+};
+
+fn do_server(sock: &mut i32, size: i32) {
+    let mut ethernet: Ethernet;
+    let mut pkt: Vec<u8> = vec![0; size as usize];
+    let mut pkt_info: PktInfo;
+    let mut recv_bytes;
+    let mut tstart = TimeSpec::zero();
+    let mut tend: TimeSpec;
+    let mut tdiff: TimeSpec;
+
+    let mut thread_handle: Option<thread::JoinHandle<()>> = None;
+
+    println!("Starting server");
+    while RUNNING.load(Ordering::Relaxed) {
+        let my_thread = thread::Builder::new().name("PrintStatsThread".to_string());
```

Comment:
Move `my_thread` outside of the loop. Keep recreating variable is not intended as I think.

Reply:
I moved it inside of the PerfReqStart match expression. Once I move my_thread out of the loop, it throws an error "use of moved value: `my_thread`". However, if I move my_thread into the match expression, it is supposed to be executed just once. here is a snippet of the code I edited
`
...
match opcode {
            PerfOpcode::PerfReqStart => {
                eprintln!("Received start '{:08x}'", id);
                let my_thread = thread::Builder::new().name("PrintStatsThread".to_string());
                ...
                }
`